packages/ckeditor5-engine/src/conversion/downcastdispatcher.js

```diff
@@ -489,6 +540,156 @@ export default class DowncastDispatcher {
 		delete this.conversionApi.consumable;
 	}
 
+	/**
+	 * Internal method for converting element insert. It will fire events for the inserted element and events for its attributes.
+	 *
+	 * @private
+	 * @fires insert
+	 * @fires attribute
+	 * @param {Object} data Event data.
+	 */
+	_convertInsertWithAttributes( data ) {
+		this._testAndFire( 'insert', data );
+
+		// Fire a separate addAttribute event for each attribute that was set on inserted items.
+		// This is important because most attributes converters will listen only to add/change/removeAttribute events.
+		// If we would not add this part, attributes on inserted nodes would not be converted.
+		for ( const key of data.item.getAttributeKeys() ) {
+			data.attributeKey = key;
+			data.attributeOldValue = null;
+			data.attributeNewValue = data.item.getAttribute( key );
+
+			this._testAndFire( `attribute:${ key }`, data );
+		}
+	}
+
+	/**
+	 * Get changes without those that needs to be converted using {@link #convertRefresh} defined by a `triggerBy` configuration for
+	 * {@link module:engine/conversion/downcasthelpers~DowncastHelpers#elementToElement `elementToElement()`} conversion helper.
+	 *
+	 * @param {module:engine/model/differ~Differ} differ The differ object with buffered changes.
+	 * @returns {Array.<Object>}
+	 * @private
+	 */
+	_getChangesAfterAutomaticRefreshing( differ ) {
+		const changes = differ.getChanges();
+		const elementsToRefresh = this._getElementsForAutomaticRefresh( changes );
+
+		if ( !elementsToRefresh.size ) {
+			return changes;
+		}
+
+		for ( const element of elementsToRefresh.values() ) {
+			differ.refreshItem( element );
+		}
+
+		// The `differ.refreshItem()` invalidates differ cache - we can't re-use previous changes.
+		const changesAfterRefresh = differ.getChanges();
+
+		return changesAfterRefresh.filter( entry => !elementsToRefresh.has( getElementFromChange( entry ) ) );
```

Comment:
Maybe a comment about why we filter-out some changes would be good.

Reply:
There's an explanation in the method jsdoc. If you think it is not clear I can add it here.

Reply:
There's an explanation in the method jsdoc. If you think it is not clear I can add it here.

Reply:
Differ does not know which elements are related by a `triggerBy` - I think that this is the responsibility of the downcast dispatcher. It can read those mappings and act upon them to alter the conversion process. Mappings are defined by `elementToElement()` conversion helper. So

1.  The `elementToElement()` defines which events should trigger reconversion.
2.  The `DowncastDispatcher` handles the conversion process (which events to fire, which change).
3.  The `Differ` shows what changed, so if an element was inserted, removed, have changed attributes, or if it was marked to be refresh.

The `triggerBy` is a declarative mechanism of reconverting the entire element

But you can use `Differ.refreshItem()` implicitly to mark an element to be re-converted (refreshed), ie in a post-fixer.

Reply:
:+1: